pandas/core/missing.py

```diff
@@ -973,3 +981,86 @@ def _rolling_window(a: npt.NDArray[np.bool_], window: int) -> npt.NDArray[np.boo
     shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)
     strides = a.strides + (a.strides[-1],)
     return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
+
+
+def _can_hold_element(values, element: Any) -> bool:
+    """
+    Expanded version of core.dtypes.cast.can_hold_element
+    """
+    element = extract_array(element, extract_numpy=True)
+    return can_hold_element(values, element)
+
+
+def fillna_ea_array(values, value, limit=None, inplace: bool = False, downcast=None):
+    """
+    Fillna logic for ExtensionArrays.
+
+    Dispatches to the EA.fillna method (in which case downcast is currently
+    ignored), except for datetime64 in which case fallback to object dtype
+    is currently allowed.
+    """
+    if not _can_hold_element(values, value) and values.dtype.kind == "M":
+        # We support filling a DatetimeTZ with a `value` whose timezone
+        #  is different by coercing to object.
+        # TODO: don't special-case td64
+        values = values.astype(object)
+        return fillna_array(values, value, limit=limit, inplace=True, downcast=downcast)
+
+    values = values if inplace else values.copy()
+    return values.fillna(value, limit=limit)
+
+
+def fillna_array(values, value, limit=None, inplace: bool = False, downcast=None):
```

Comment:
could be used for Block.fillna too?

Reply:
I don't directly see an easy way to do that. The Block.fillna additionally has the "split" logic, which I personally don't want to put in here. And it's not directly straightforward to split a part of Block.fillna. 

I could for example let this function return some sentinel like `NotImplemented` or None to signal to Block.fillna that the filling hasn't been done yet, and it should split the block and call fillna again. In which case most of the splitting logic stays in the internals. I could try something like this if you want, but it's also a bit of a hack.

Reply:
> The Block.fillna additionally has the "split" logic, which I personally don't want to put in here

100% agree

I've been looking at simplifying the downcasting in a way that would make the splitting much easier to separate the splitting from non-splitting (not just for fillna) (xref #44241).  It would be an API change so wouldn't be implement in time to simplify this (and your putmask PR), but would eventually allow for some cleanup.